File: ecommerce/ecommerce-mini/src/services/coupon.js
import { calculateDiscount } from '../domain/logic.js'

export class CouponService {
  constructor(couponRepo) {
    this.couponRepo = couponRepo
  }

  validate(couponId, orderTotalCents) {
    const coupon = this.couponRepo.findById(couponId)
    if (!coupon) {
      throw new Error('COUPON_NOT_FOUND')
    }

    if (coupon.status !== 'UNUSED') {
      throw new Error('COUPON_ALREADY_USED')
    }

    if (orderTotalCents < coupon.minSpendCents) {
      throw new Error('COUPON_THRESHOLD_NOT_MET')
    }
    
    return coupon
  }

  /**
   * 获取最优优惠券
   * 候选集 = 全场通用券 (userId 为 null) + 该用户持有的券
   * @param {string} userId 用户ID
   * @param {number} orderTotalCents 订单总额
   * @returns {import("../domain/types.js").Coupon | null} 最优优惠券
   */
  getBestCoupon(userId, orderTotalCents) {
    const allCoupons = this.couponRepo.findAll()
    const availableCoupons = allCoupons.filter(c =>
      c.status === 'UNUSED'
      && orderTotalCents >= c.minSpendCents
      && ((c.userId ?? null) === null || c.userId === userId)
    )

    if (availableCoupons.length === 0) return null

    // 计算减免金额并排序
    const ratedCoupons = availableCoupons.map(coupon => ({
      coupon,
      discount: calculateDiscount(orderTotalCents, coupon)
    }))

    // 按折扣额降序，若相同则按 ID 升序以保证确定性
    ratedCoupons.sort((a, b) => {
      if (b.discount !== a.discount) return b.discount - a.discount
      return a.coupon.id.localeCompare(b.coupon.id)
    })

    return ratedCoupons[0].coupon
  }

  redeem(couponId) {
    const coupon = this.couponRepo.findById(couponId)
    if (coupon) {
      coupon.status = 'USED'
      this.couponRepo.save(coupon)
    }
  }

  /**
   * 列出当前用户可见的券：全场通用券 (ACTIVE 规则模板除外) + 本人持有的券
   * @param {string} [userId] 不传时仅返回全场通用券
   */
  list(userId = null) {
    return this.couponRepo.findAll().filter(c => {
      const ownerId = c.userId ?? null
      if (ownerId !== null) return ownerId === userId
      return c.status !== 'ACTIVE'
    })
  }
}

File: ecommerce/ecommerce-mini-python/src/services/coupon.py
from typing import List, Optional
from ..domain.models import Coupon, calculate_discount
from ..repo.memory import MemoryRepo

class CouponService:
    def __init__(self, repo: MemoryRepo[Coupon]):
        self.repo = repo

    def validate(self, coupon_id: str, order_total_cents: int) -> Coupon:
        coupon = self.repo.find_by_id(coupon_id)
        if not coupon:
            raise ValueError("COUPON_NOT_FOUND")

        if coupon.status != "UNUSED":
            raise ValueError("COUPON_ALREADY_USED")

        if order_total_cents < coupon.min_spend_cents:
            raise ValueError("COUPON_THRESHOLD_NOT_MET")

        return coupon

    def get_best_coupon(self, user_id: str, order_total_cents: int) -> Optional[Coupon]:
        # 候选集 = 全场通用券 (user_id 为 None) + 该用户持有的券
        all_coupons = self.repo.find_all()
        available_coupons = [
            c for c in all_coupons
            if c.status == "UNUSED"
            and order_total_cents >= c.min_spend_cents
            and (c.user_id is None or c.user_id == user_id)
        ]

        if not available_coupons:
            return None

        # 计算减免额并排序
        rated = []
        for coupon in available_coupons:
            discount = calculate_discount(order_total_cents, coupon)
            rated.append((discount, coupon))
        
        # 按折扣额降序，相同则按 ID 升序
        rated.sort(key=lambda x: (-x[0], x[1].id))
        
        return rated[0][1]

    def redeem(self, coupon_id: str):
        coupon = self.repo.find_by_id(coupon_id)
        if coupon:
            coupon.status = "USED"
            self.repo.save(coupon.id, coupon)

    def list_coupons(self, user_id: Optional[str] = None) -> List[Coupon]:
        # 当前用户可见的券：全场通用券 (ACTIVE 规则模板除外) + 本人持有的券
        result = []
        for c in self.repo.find_all():
            if c.user_id is not None:
                if c.user_id == user_id:
                    result.append(c)
            elif c.status != "ACTIVE":
                result.append(c)
        return result
